bn_mp_radix_size_overestimate.c

```diff
@@ -0,0 +1,254 @@
+    9008676367646627443ull,  9223372036854775808ull,  9425047617917838031ull,
+    9615192413521158631ull,  9795053960520401544ull,  9965687683920283007ull,
+   10127994423561913434ull, 10282749222901319551ull, 10430623713119086168ull,
+   10572203729794814195ull, 10708003330985790206ull, 10838476064306023373ull,
+   10964024106461197019ull, 11085005739835568998ull, 11201741516507896395ull,
+   11314519376860321395ull, 11423598928577852074ull, 11529215046068469760ull,
+   11631580915841357939ull, 11730890627131531983ull, 11827321386901076197ull,
+   11921035422734852583ull, 12012181624960263419ull, 12100896969734095496ull,
+   12187307757246061761ull, 12271530693133976959ull, 12353673836347420459ull,
+   12433837432775607386ull, 12512114650772492370ull, 12588592232115013503ull,
+   12663351069800359783ull, 12736466722332780120ull, 12808009872697200099ull,
+   12878046739008508147ull, 12946639442816362188ull, 13013846340199484158ull,
+   13079722320071570371ull, 13144319073519717325ull, 13207685337486676934ull,
+   13269867115674890971ull, 13330907879180520702ull, 13390848749049262950ull,
+   13449728662674133884ull, 13507584525721590347ull, 13564451351070528819ull,
+   13620362386074015347ull, 13675349229302017118ull, 13729441937791546026ull,
+   13782669125715645774ull, 13835058055282163712ull
+};
+#else
+#  ifdef _MSC_VER
+#    pragma message("mp_radix_size_overestimate: unknown or undefined MP_xBIT type")
+#  else
+#    warning "mp_radix_size_overestimate: unknown or undefined MP_xBIT type"
+#  endif
+#endif
+/* *INDENT-ON* */
+
+mp_err mp_radix_size_overestimate(const mp_int *a, const int base, int *size)
+{
+   mp_int bi_bit_count, bi_k;
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   mp_int bi_k_bis;
+#endif
+   int rem, bit_count;
+   mp_err e = MP_OKAY;
+
+   if ((base < 2) || (base > 64)) {
+      *size = 0;
+      return MP_VAL;
+   }
+   /* floor( log_2(a) ) + 1 */
+   bit_count = mp_count_bits(a);
+
+   if (bit_count == 0) {
+      *size = 2;
+      return MP_OKAY;
+   }
+
+   /* A small shortcut for powers of two. */
+   if (!(base&(base-1))) {
+      unsigned int x = (unsigned int)base;
+      int y;
+      for (y=0; (y < 7) && !(x & 1u); y++) {
+         x >>= 1;
+      }
+      *size = bit_count/y;
+      rem = bit_count - ((*size) * y);
+      /* Add 1 for the remainder if any and 1 for "\0". */
+      *size += (rem == 0) ? 1 : 2;
+      /* And one extra character for the minus sign */
+      if (a->sign == MP_NEG) {
+         (*size)++;
+      }
+      return MP_OKAY;
+   }
+
+   /* d(bitcount,base) = floor( (bitcount * 2^p) / k(base) ) + 1 */
+
+   if ((e = mp_init_multi(&bi_bit_count, &bi_k, NULL)) != MP_OKAY) {
+      *size = 0;
+      return e;
+   }
+#if ( (defined MP_8BIT) || (defined MP_16BIT) )
+   if ((e = mp_init(&bi_k_bis)) != MP_OKAY) {
+      /* No "goto" to avoid cluttering this code with even more preprocessor branches */
+      mp_clear_multi(&bi_bit_count, &bi_k, NULL);
+      *size = 0;
```

Comment:
I just asked myself if 0 is really the good value to return on error... whatever we choose, for sure we should set it once on function-entry and not x-times on each error-case

Reply:
> I just asked myself if 0 is really the good value to return on error

The returns on error are `MP_VAL`s?
Yes, we could set `size` to `-1` but as we want to change the type of all size-related variables to `size_t` later we shouldn't introduce things that **we know** get changed again later.

And also: if the user skips the error-check and use the resulting `size` for `malloc` directly and if that value is `-1` and it is a 32 bit machine with `sizeof(size_t) == 4` and there is enough RAM and and/or space in the swap-partition/file&hellip;

> set it once on function-entry

Yepp, that's right, will do.

Ahm, now that you mentioned it, would it be a good idea to check if we have something to write on in the first place? I  mean `if (size == NULL) ...`?

Reply:
> And also: if the user skips the error-check 

that was the origin of my thought...

> and use the resulting `size` for `malloc` directly and if that value is `-1` and it is a 32 bit machine with `sizeof(size_t) == 4` and there is enough RAM and and/or space in the swap-partition/file…

:laughing: 

> Ahm, now that you mentioned it, would it be a good idea to check if we have something to write on in the first place? I mean `if (size == NULL) ...`?

puh, that's nearly nowhere checked... I think yes, we should do it, but maybe we should first discuss it?

Reply:
> but maybe we should first discuss it?

Hu? Oh, left it in after testing, sorry.